vulyk/blueprints/gamification/core/rules.py

```diff
@@ -0,0 +1,258 @@
+# -*- coding: utf-8 -*-
+"""
+Classes within the package are supposed to be intermittent containers between
+JSON sources and final queries. Everything that has anything to do with images,
+representation and other specifications also goes here.
+"""
+__all__ = [
+    'Rule',
+    'ProjectRule',
+    'RuleValidationException'
+]
+
+
+class RuleValidationException(Exception):
+    pass
+
+
+class Rule:
+    """
+    Base class for all rule containers.
+    Those rules may be interpreted like the following:
+
+    - you closed n tasks
+    - you closed n tasks in m days
+    - you closed n tasks in mornings in m days
+    - you closed n tasks in weekends
+    - you've been working for m weekends
+    - you've been working for m days in a row
+    - you've been working for m weekends in a row
+
+    Also additional bonuses to be given after user gets the achievement are
+    included as well as achievement and its badge is.
+    """
+
+    def __init__(self,
+                 id: int,
+                 badge: str,
+                 name: str,
+                 description: str,
+                 bonus: int,
+                 tasks_number: int,
+                 days_number: int,
+                 is_weekend: bool,
+                 is_adjacent: bool) -> None:
+        """
+        :param id: Raw JSON representation hash.
+        :type id: int
+        :param badge: Badge image (either base64 or URL)
+        :type badge: str
+        :param name: Achievement name
+        :type name: str
+        :param description: Achievement textual description
+        :type description: str
+        :param bonus: How much additional points we give for the achievement
+        :type bonus: int
+        :param tasks_number: Number of tasks needed to achieve
+        :type tasks_number: int
+        :param days_number: Number of days needed to achieve
+        :type days_number: int
+        :param is_weekend: Should this achievement be tied up with weekends?
+        :type is_weekend: bool
+        :param is_adjacent: Must it be given for work in adjacent days?
+        :type is_adjacent: bool
+        """
+        self.badge = badge
+        self.name = name
+        self.description = description
+        self.bonus = bonus
+
+        self._tasks_number = tasks_number
+        self._days_number = days_number
+        self._is_weekend = is_weekend
+        self._is_adjacent = is_adjacent
+
+        self._hash = id
+
+        self._validate()
+
+    def _validate(self):
+        """
```

Comment:
Maybe one day you should change that into list of tiny helpers, which you can iterate:

```self._validators = [validate_adj, validate_whatever]

def _validate(self):
 for v in self._validators:....
```